### PR TITLE
Calculate reference dates

### PR DESCRIPTION
**Pull request by bitdivision**
This is currently blocked on #1731 (Which may cause significant changes)

### What is the context of this PR?
We need to add a few reference date ranges, which all amount to dates which are a constant offset from the previous Sunday (not including today if it is Sunday).

There is additional functionality in the formatting of these date ranges. If the range spans a single month and year, then the first month / year should not be shown. e.g. `Thursday 16 to Friday 17 August 2018` rather than `Thursday 16 August 2018 to Friday 17 August 2018`. I've assumed this change is required going from the examples, but let me know if that's incorrect.

### How to review 

* The test schema contains a few examples as per the examples we were given. If testing manually with go-launcher, remove `started_at` from metadata to default to `utcnow()`.
* Ensure these examples match the expected ranges from trello.

### Notable changes
Look over the following code in particular:

* `started_at` changed from `now(timezone.utc)` to `utcnow()`. Due to discrepancies between these, I think we should standardise `utcnow()`.
* `metadata_parser.py` there is a special case here for validating started_at for the first question when it isn't present. This also defaults to `utcnow()` . See more detailed comment in trello.
* I've added a dependency on `freezegun` in order to pin `started_at` after the first question has been answered in integration tests. I can't see a way to inject any metadata for the second question, but if anyone knows of a way to do this nicely I'd love to remove this.

I'll need to setup signed commits before this is merged.

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -50,6 +50,7 @@ blinker = "*"
 humanize = "*"
 flask-talisman = "*"
 marshmallow = "*"
+freezegun = "*"
```
Comment (jonnyshaw89) [inline Pipfile]: shouldn't this be in `[dev-packages]`?

File: app/jinja_filters.py [hunk 2]
```diff
@@ -142,6 +150,43 @@ def format_conditional_date(context, date1=None, date2=None):
...
+    :param (int) offset_days: The number of days to offset this date`
+    :param (str) day_of_week: The day of the previous week to offset from (two letter abbreviation)
+    :returns (str): The offset date
+    """
+    if 'T' in input_datetime:
+        if '.' in input_datetime:
+            parsed_datetime = datetime.strptime(input_datetime, '%Y-%m-%dT%H:%M:%S.%f')
+        else:
+            # Freezegun doesn't support microseconds.
+            parsed_datetime = datetime.strptime(input_datetime, '%Y-%m-%dT%H:%M:%S')
```
Comment (jonnyshaw89) [inline app/jinja_filters.py]: is this just here to support tests?

File: app/submitter/converter.py [hunk 3]
```diff
@@ -46,7 +46,7 @@ def convert_answers(metadata, schema, answer_store, routing_path, flushed=False)
       }
     """
     survey_id = schema.json['survey_id']
-    submitted_at = datetime.now(timezone.utc)
+    submitted_at = datetime.utcnow()
```
Comment (jonnyshaw89) [inline app/submitter/converter.py:50]: 👍

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: Pipfile (hunk 1 above)
Comment (bitdivision) [reply Pipfile]: Yikes. Now sure how that happened.

File: app/jinja_filters.py (hunk 2 above)
Comment (bitdivision) [reply app/jinja_filters.py]: Yes, and it's horrendous. I'd ideally like to get rid of freezegun, but couldn't see a way to inject metadata here.

**Comment by bitdivision** [comment]
Now that #1731 has been sorted out, this should be ready to go in.

File: app/jinja_filters.py [hunk 4]
```diff
@@ -142,6 +152,42 @@ def format_conditional_date(context, date1=None, date2=None):
...
+    :param (int) offset_days: The number of days to offset this date`
+    :param (str) day_of_week: The day of the previous week to offset from (two letter abbreviation)
+    :returns (str): The offset date
+    """
+    if input_datetime:
+        if 'T' in input_datetime:
+            parsed_datetime = datetime.strptime(input_datetime, '%Y-%m-%dT%H:%M:%S.%f')
+        else:
+            parsed_datetime = datetime.strptime(input_datetime, '%Y-%m-%d')
+    else:
```
Comment (dwyeradam) [inline app/jinja_filters.py]: Not sure I like this. Possibly use result of `input_datetime.split("T")[0]` instead.
```
if input_datetime:
   parsed_datetime = datetime.strptime(input_datetime.split("T")[0], '%Y-%m-%d')
else:
```

**Review by dwyeradam — CHANGES_REQUESTED** [review]

File: app/storage/metadata_parser.py [hunk 5]
```diff
@@ -92,6 +92,9 @@ def _validate_metadata_is_present(metadata, required_metadata):
         if name == 'trad_as_or_ru_name':
             # either of 'trad_as' or 'ru_name' is required
             valid = bool(metadata.get('trad_as') or metadata.get('ru_name'))
+        elif name == 'started_at':
+            # started_at will not normally be passed to us, so we can't really validate immediately.
+            valid = True
```
Comment (bitdivision) [inline app/storage/metadata_parser.py]: I previously had a line here:
```	
            if not metadata.get('started_at'): 
                metadata['started_at'] = datetime.utcnow().isoformat()
```
I don't think this is needed anymore, but please correct me if you think it is.

File: app/jinja_filters.py [hunk 6]
```diff
@@ -142,6 +152,39 @@ def format_conditional_date(context, date1=None, date2=None):
     return format_date(context, date)
 
 
+@blueprint.app_template_filter()
+def calculate_offset_date(input_datetime, offset_days, day_of_week='SU'):
```
Comment (dwyeradam) [inline app/jinja_filters.py]: I've changed my mind. As initially discussed I think this should accept a dictionary rather than only support days.
So along the lines of:
```
calculate_offset_date(input_datetime, offset, day_of_week='SU'):

   offset_days = offset.get('days', 0)
   offset_weeks = offset.get('weeks', 0)
   offset_years = offset.get('years', 0)
```

The filter would then look like this:
`calculate_offset_date(collection_metadata['started_at'], {'weeks'-4, 'days':1})`

**Review by dwyeradam — CHANGES_REQUESTED** [review]

**Review by dwyeradam — CHANGES_REQUESTED** [review]
We also need a change in https://github.com/ONSdigital/go-launch-a-survey
At the moment if someone launches this with `"started_at"` in the metadata (only a problem with our launcher) it will 500. This should default to either blank or a useable date format.

File: app/jinja_filters.py (hunk 6 above)
Comment (bitdivision) [reply app/jinja_filters.py]: Fixed now.

**Comment by bitdivision** [comment]
Good point about `started_at` in the launcher. We could actually remove it from metadata in that schema since it is now in collection_metadata, but I think it is useful for the purposes of testing.

Will make a PR for go-launcher to sort this out @dwyeradam

**Comment by bitdivision** [comment]
I've changed the default day of week to Monday, and updated the test schema to match as per @dwyeradam's suggestion. By using the Monday most of the time, and changing to Sunday where necessary, we avoid using strange numbers in the offset.

**Comment by bitdivision** [comment]
One more change has gone in to remove the handling of in-flight `started_at` in metadata. 

This is because the `started_at` in production currently is defective, and defaulting to now is not going to make any real difference. Additionally, the questionnaires that use `started_at` are not currently in production.

Since `started_at` has been removed from the schema, this PR becomes more difficult to test manually, and the unit / integration tests should be testing all edge cases.

**Comment by bitdivision** [comment]
@AndrewTorrance I've just double checked that babel does format days of the week in welsh:
```
>>> format_datetime(d, locale='cy', format='full')
'Dydd Iau, 23 Awst 2018 am 13:42:12 Amser Cyffredniol Cydlynol'
>>> format_datetime(d, locale='en', format='full')
'Thursday, August 23, 2018 at 1:42:12 PM Coordinated Universal Time'
```
I assume that's correct.

I can try to add a test for this if you think its needed, but I'm not sure it's really worth it?

**Comment by bitdivision** [comment]
Need to make a change to schema validator to fix travis.

File: tests/integration/mci/test_date_reference_ranges.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
+from freezegun import freeze_time
+from tests.integration.integration_test_case import IntegrationTestCase
+
+@freeze_time('2018-10-08')
```
Comment (dwyeradam) [inline tests/integration/mci/test_date_reference_ranges.py:4]: I've never used this but I will in future 👍.

**Review by dwyeradam — CHANGES_REQUESTED** [review]
Functional Tests?

File: app/jinja_filters.py [hunk 8]
```diff
@@ -142,6 +152,45 @@ def format_conditional_date(context, date1=None, date2=None):
     return format_date(context, date)
 
 
+@blueprint.app_template_filter()
+def calculate_offset_date(input_datetime, offset, day_of_week='MO'):
```
Comment (dwyeradam) [inline app/jinja_filters.py]: This is really hard to name but I think we may need to be clearer about the fact that this offset starts from a specified day in the last whole week. We'll need to buy a hat to put suggestions in.

**Comment by bitdivision** [comment]
Functional tests present issues in terms of mocking started_at now that we've removed it from metadata.

I'll have a think about potential solutions, but I'd really rather avoid reproducing the offset code in the tests.

File: app/jinja_filters.py (hunk 8 above)
Comment (bitdivision) [reply app/jinja_filters.py]: `calculate_date_offset_from_last_weekday` - I think that works?

**Comment by bitdivision** [comment]
The only obvious way I can see to mock collection_metadata for functional tests is to pick it up from the jwt if it's available.

This would still be adding code that's only really used in the tests though. @jonnyshaw89 @dwyeradam any thoughts?

File: app/helpers/template_helper.py [hunk 9]
```diff
@@ -34,6 +34,17 @@ def session_wrapper(*args, **kwargs):
     return session_wrapper
 
 
+def with_collection_metadata_context(func):
+    @wraps(wraps)
+    def collection_metadata_context_wrapper(*args, **kwargs):
```
Comment (AndrewTorrance) [inline app/helpers/template_helper.py]: Looks to be duplicating existing decorator ?

**Review by AndrewTorrance — COMMENTED** [review]
Looks good , one question regarding the new decorator. Not putting it in as a blocker as I will not be here to take it off next week.

**Comment by bitdivision** [comment]
Absolutely right about the decorator. I've removed it entirely now.

**Comment by bitdivision** [comment]
@jonnyshaw89 @dwyeradam I'm fairly confident that this is finally ready to go in :)

File: tests/functional/spec/date_reference_ranges.spec.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  // We have no way to mock out started_at, the unit and integration tests cover the date values
+  // this test is only to validate the general questionnaire functionality and to make sure
+  // the date format is correct.
+  it('Given we start the survey, when we click the button, the dates should be in the correct format...', function() {
+    return helpers.startQuestionnaire('test_date_reference_ranges.json').then(() => {
+        return browser
+          // Should show a date range in the right format
+
+          // Matches `in the week Friday 24 August 2018 to`. The day of the month can be 1 or two characters
+          .getText(SummaryPage.questionText()).should.eventually.match(new RegExp('in the week [a-zA-Z]+ \\d(?:\\d)? [a-zA-Z]+ \\d{4} to'))
```
Comment (jonnyshaw89) [inline tests/functional/spec/date_reference_ranges.spec.js:18]: 👍

File: data/en/test_date_reference_ranges.json [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    }, {
+        "name": "ru_name",
+        "validator": "string"
+    }],
+    "sections": [{
+        "id": "default-section",
+        "groups": [{
+            "blocks": [{
+                "type": "Introduction",
+                "id": "introduction",
```
Comment (jonnyshaw89) [inline data/en/test_date_reference_ranges.json]: I would probably drop the introduction from this test schema?

**Review by jonnyshaw89 — APPROVED** [review]
Looks good. I would just drop the intro block from the test schemas as I cant see it adds anything.

File: tests/functional/spec/date_reference_ranges.spec.js [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+const DateSeparateBlockPage = require('../pages/features/date_reference_ranges/date-separate-block.page.js');
+const DateRangeBlockPage = require('../pages/features/date_reference_ranges/date-range-block.page.js');
+const SummaryPage = require('../pages/features/date_reference_ranges/summary.page.js');
+
+describe('Reference Date Range Checks', function() {
+  // We have no way to mock out started_at, the unit and integration tests cover the date values
+  // this test is only to validate the general questionnaire functionality and to make sure
+  // the date format is correct.
+  it('Given we start the survey, when we click the button, the dates should be in the correct format...', function() {
+    return helpers.startQuestionnaire('test_date_reference_ranges.json').then(() => {
```
Comment (jonnyshaw89) [inline tests/functional/spec/date_reference_ranges.spec.js]: I think you need to change this to `openQuestionnaire` as it doesn't have intro.

Comment (bitdivision) [reply tests/functional/spec/date_reference_ranges.spec.js]: I did run the tests after the change, and it did work. I'll switch it over though.

Comment (bitdivision) [reply tests/functional/spec/date_reference_ranges.spec.js]: Apparently I didn't :)

**Review by dwyeradam — APPROVED** [review]
